File: content/problems/0005-longest-palindromic-substring/solutions/expand-around-center/Solution.kt
class Solution {
    fun longestPalindrome(s: String): String {
        var bestLeft = 0
        var bestRight = 0
        for (center in s.indices) {
            val length = maxOf(expand(s, center, center), expand(s, center, center + 1))
            if (length > bestRight - bestLeft + 1) {
                bestLeft = center - (length - 1) / 2
                bestRight = center + length / 2
            }
        }
        return s.substring(bestLeft, bestRight + 1)
    }

    private fun expand(s: String, startLeft: Int, startRight: Int): Int {
        var left = startLeft
        var right = startRight
        while (left >= 0 && right < s.length && s[left] == s[right]) {
            left--
            right++
        }
        return right - left - 1
    }
}

File: content/problems/0005-longest-palindromic-substring/solutions/expand-around-center/solution.cpp
class Solution {
    int expand(const string& s, int left, int right) {
        while (left >= 0 && right < (int)s.size() && s[left] == s[right]) {
            --left;
            ++right;
        }
        return right - left - 1;
    }
public:
    string longestPalindrome(string s) {
        int bestLeft = 0, bestRight = 0;
        for (int center = 0; center < (int)s.size(); ++center) {
            int length = max(expand(s, center, center), expand(s, center, center + 1));
            if (length > bestRight - bestLeft + 1) {
                bestLeft = center - (length - 1) / 2;
                bestRight = center + length / 2;
            }
        }
        return s.substr(bestLeft, bestRight - bestLeft + 1);
    }
};
